src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java

```diff
@@ -72,15 +75,91 @@ public SonarBuildWrapper(@Nullable String installationName) {
   }
 
   @Override
-  public DescriptorImpl getDescriptor() {
-    return (DescriptorImpl) super.getDescriptor();
+  public void setUp(Context context, Run<?, ?> build, FilePath workspace, Launcher launcher, TaskListener listener, EnvVars initialEnvironment)
+    throws IOException, InterruptedException {
+
+    if (!SonarInstallation.isValid(getInstallationName(), listener)) {
+      return;
+    }
+
+    SonarInstallation installation = SonarInstallation.get(getInstallationName());
+
+    String msg = Messages.SonarBuildWrapper_Injecting(installation.getName());
+    Logger.LOG.info(msg);
+    listener.getLogger().println(msg);
+
+    context.getEnv().putAll(createVars(installation));
+
+    context.setDisposer(new AddBuildInfo(installation));
+
+    build.addAction(new SonarMarkerAction());
+  }
+
+  @VisibleForTesting
+  static Map<String, String> createVars(SonarInstallation inst) {
+    Map<String, String> map = new HashMap<>();
+
+    map.put("SONAR_CONFIG_NAME", inst.getName());
+    String hostUrl = getOrDefault(inst.getServerUrl(), "http://localhost:9000");
+    map.put("SONAR_HOST_URL", hostUrl);
+    String login = getOrDefault(inst.getSonarLogin(), "");
+    map.put("SONAR_LOGIN", login);
+    String password = getOrDefault(inst.getSonarPassword(), "");
+    map.put("SONAR_PASSWORD", password);
+    String token = getOrDefault(inst.getServerAuthenticationToken(), "");
+    map.put("SONAR_AUTH_TOKEN", token);
+
+    map.put("SONAR_JDBC_URL", getOrDefault(inst.getDatabaseUrl(), ""));
+
+    String jdbcDefault = SQServerVersions.SQ_5_1_OR_LOWER.equals(inst.getServerVersion()) ? DEFAULT_SONAR : "";
+    map.put("SONAR_JDBC_USERNAME", getOrDefault(inst.getDatabaseLogin(), jdbcDefault));
+    map.put("SONAR_JDBC_PASSWORD", getOrDefault(inst.getDatabasePassword(), jdbcDefault));
+
+    if (StringUtils.isEmpty(inst.getMojoVersion())) {
+      map.put("SONAR_MAVEN_GOAL", "sonar:sonar");
+    } else {
+      map.put("SONAR_MAVEN_GOAL", SonarUtils.getMavenGoal(inst.getMojoVersion()));
+    }
+
+    map.put("SONAR_EXTRA_PROPS", getOrDefault(getAdditionalProps(inst), ""));
+
+    StringBuilder sb = new StringBuilder();
+    sb.append("{ \"sonar.host.url\" : \"").append(StringEscapeUtils.escapeJson(hostUrl)).append("\"");
+    if (!login.isEmpty() || !token.isEmpty()) {
+      sb.append(", \"sonar.login\" : \"").append(StringEscapeUtils.escapeJson(token.isEmpty() ? login : token)).append("\"");
+    }
+    if (!password.isEmpty()) {
+      sb.append(", \"sonar.password\" : \"").append(StringEscapeUtils.escapeJson(password)).append("\"");
+    }
+    sb.append("}");
+
+    map.put("SONARQUBE_SCANNER_PARAMS", sb.toString());
+
+    // resolve variables against each other
+    Map<String, String> sonarEnvResolved = new HashMap<>(map);
+    EnvVars.resolve(sonarEnvResolved);
+
+    return sonarEnvResolved;
```

Comment:
Just out of curiosity, why do we need a new map here? Can't you just `EnvVars.resolve(map)` and `return map`?